### PR TITLE
Adds Noah-MP precipitation timestep fix 

### PR DESCRIPTION
**Pull request by TimLahmers**
Adds Noah-MP precipitation timestep fix for cases where LIS dt is greater than forcing dt

Description

Corrects issue #710

This pull request includes an update to Noah-MP v4.0.1. The lines to compute the precipitation assume that the Noah-MP timestep is less than or equal to the forcing timestep. In the event that the Noah-MP timestep exceeds the forcing timestep (i.e. is longer), forcing precipitation is multiplied by a shorter timestep and is therefore reduced.

This PR addresses that issue with an added if-statement that compares the model and forcing timestep. If the forcing timestep is shorter, then a temporary timestep (tmp_dt) is assigned to the model timestep and used and used for this calculation. If the condition is not true, the calculation used in previous version of Noah-MP remains valid.

Testcase

The benefits of this case are demonstrated in the directories:
/discover/nobackup/projects/nu-wrf/members/tlahmers/VALID_RUNS/LISF_precip_valid/run_ctrl
/discover/nobackup/projects/nu-wrf/members/tlahmers/VALID_RUNS/LISF_precip_valid/run_update

In the OUTPUT directories of the control simulation (run_ctrl), the error is issue #710 is demonstrated, as the Rainf_tavg and Snowf_tavg variables are half of their actual values. In this case, the Noah-MP timestep is 60 minutes and the IMERG forcing timestep is 30 minutes. In a case with all rain precipitation, this is easy to demonstrate as Rainf_tavg is half of the TotalPrecip_tavg variable due to the aforementioned logic error.

In the updated simulation (run_update), these errors are corrected, as Rainf_tavg (and Snowf_tavg) are double their values from the control simulation. For example, Rainf_tavg would be equivalent to TotalPrecip_tavg for cases of all rain.

This change is necessary for some model configurations where a long Noah-MP timestep is used with a higher resolution precipitation product (e.g. LIS/WRF-Hydro).

### CODE REVIEW

**Comment by TimLahmers** [comment]
@bmcandr Thanks for your guidance on updating my pull request. For my own future reference, do I use the latest public release branch for bug fixes and the master branch for new features? Also, do you need me to re-run my test cases since they are based on the master branch?

**Comment by bmcandr** [comment]
You're welcome. Generally:
* master: new features (and fixes for any bugs found in those new features) and major changes
* support/lisf-public-7.3: bug fixes for code in the latest release

See [Working with GitHub 4.6](https://github.com/NASA-LIS/LISF/blob/master/docs/working_with_github/working_with_github.adoc#46-step-4-carry-out-your-work) for more. The support branch is periodically merged back into master to propagate the fixes.

I'll try running the testcases as is and if the results look off I'll let you know.

**Comment by TimLahmers** [comment]
@bmcandr Thanks for the updates. This update is important for LIS/WRF-Hydro stakeholders, so please let me know if there is any way I can help to expedite this process.

**Comment by bmcandr** [comment]
I am waiting for my test run to kick off. In the meantime, @dmocko does this change look good to you?

**Comment by bmcandr** [comment]
@TimLahmers, I successfully reproduced your output. Can you confirm David's condition that "the accumulated precip in the output matches what would be expected" is met?

**Comment by TimLahmers** [comment]
@bmcandr Do you want me to sum the snow and rain in the model to confirm it matches total precipitation? I can write some code to do that this afternoon.

**Comment by bmcandr** [comment]
That sounds good @TimLahmers. Thanks!

**Comment by TimLahmers** [comment]
@bmcandr Earlier this week I confirmed that precipitation errors after the fix were generally several orders of magnitude smaller than precipitation (I attribute them to round-off errors). I did this evaluation with the prior code base (i.e. master). Is that sufficient, or should I also check on the support/lisf-public-7.3 branch?

**Comment by dmocko** [comment]
Hi @TimLahmers .  I would like to confirm the following: If you compare the LIS output Rainf_f (and TotalPrecip), does that match the precipitation amounts in the metforcing original files (outside of LIS)?  This should be true for both the case when the model timestep is less than forcing timestep (which is usually the case, and we have confirmed that the amounts match) - as well as now when the model timestep is greater than the forcing timestep.

For example, a monthly-average (or maybe even a daily average) could confirm this.  I don't think we need a testcase to confirm it, but I would like to hear from you that you can confirm this.  Maybe you already have.

**Comment by TimLahmers** [comment]
I ran the Noah-MP LSM with both 60-min and 30-min time steps, and there are still some slight errors. They are orders of magnitude less than the quantities of rain and snow and seem to be related to differences in precipitation type between time steps (rain errors are opposite snow errors). Sample plots from 03-15-2019 are attached. Do you want me to show more analysis for this?
03152019rainErr
03152019snowErr
03152019rain
03152019snow

**Comment by dmocko** [comment]
Hi @TimLahmers - thanks for sharing these.  I'm not concerned about the rain/snow errors, as you note the temperatures will be different with different timesteps, thus the rain/snow determination will be different.

I'm not sure what units you are showing.  I'm assuming kg m-2 sec-1.  What exactly are the first two panels?  Differences of what?

The fact that the differences are generally centered at zero is encouraging, but I might hope for the error differences to be more random patterns rather than their current appearance.

I think I need to convince myself, so I will do a few tests with this code and confirm that the precipitation totals will remain the same over a longer period.

**Comment by TimLahmers** [comment]
@dmocko Yes, units are kg m-2 sec-1. First two panels are differences of rain and snow, respectively. 

If you want to do more analysis, my run directory is: /discover/nobackup/projects/nu-wrf/members/tlahmers/VALID_RUNS/LISF_precip_valid/run_update/

The output directories or 30 and 60 minute simulations are:
OUTPUT.30mn
OUTPUT.60mn

The corresponding lis.config files are:
lis.config.10km.30mn.ol.rst
lis.config.10km.ol.rst    (60 minute)

**Comment by dmocko** [comment]
Hi, Thanks for the info.  I'm still testing this and hope to have an update tomorrow.

**Comment by TimLahmers** [comment]
@dmocko Have you been able to complete your testing yet? This change is necessary for a project deliverable.

Also @sujayvkumar @bmcandr I learned today that this change will also need to be pushed to the master branch, as the public branch does not include the NUOPC caps that are needed in LIS/WRF-Hydro

**Comment by bmcandr** [comment]
@TimLahmers, @jvgeiger handles merging the support branch into master. Tagged him here so he's aware.

**Comment by dmocko** [comment]
Hi,

I have tested this PR, and approve it.  Please see the attachment for details of the testing.  I have confirmed that precipitation amounts were not correct before this PR when the model timestep is greater than the forcing timestep - and that the PR fixes this bug.

I also looked at other LSMs, and I don't think this will be an issue, including Noah-MP-3.6.  This bug will only affect Noah-MP-4.0.1, which expects a precipitation amount instead of a precipitation rate.

David

[Noah-MP_timestep_precip.pdf](https://github.com/NASA-LIS/LISF/files/6474886/Noah-MP_timestep_precip.pdf)

**Comment by TimLahmers** [comment]
Thanks @dmocko for following up on this PR.

Does it just need to be approved by the LIS team now?

**Comment by jvgeiger** [comment]
Sorry, I am confused by a comment in the code.

```
   296             ! prcp: total precipitation (rainfall+snowfall)
   297             ! Both NoahMP-3.6.1 and NoahMP-4.0.1 requires total precipitation as forcing input.
   298             ! In LIS/NoahMP-3.6.1, the input forcing is total precipitation [mm], but in
   299             ! LIS/NoahMP-4.0.1, the forcing data provides precipitation rate [mm/s] !!!
```

The metforcing module provides precipitation as a rate.  I do not see anything in NoahMP 3.6 that converted this to an amount.  Is this comment wrong or is NoahMP 3.6 also broken?

**Comment by TimLahmers** [comment]
I am not familiar with this forcing difference, although I do not use Noah-MP v3.6 very often. @sujayvkumar Have you seen this issue before?

**Comment by dmocko** [comment]
Thanks @jvgeiger for your careful review of the code.

Those comments in the NoahMP401_main.F90 code are incorrect.

Noah-MP-3.6 does not convert the precipitation into an amount.  NoahMP36_main.F90 keeps the precipitation from the metforcing module as a rate.  It calls noahmp_driver_36.F90, keeping the precipitation as a rate.  And this calls module_sf_noahmplsm_36.F90, which expects the input precipitation to be a rate.

Noah-MP-4.0.1 converts the precipitation from a rate into an amount in NoahMP401_main.F90.  It calls noahmp_driver_401.F90, keeping the precipitation as an amount.  For the Noah-MP-4.0.1 implementation, we kept an additional subroutine from HRLDAS/WRF: module_sf_noahmpdrv_401.F90, which is called by noahmp_driver_401.F90.  The module_sf_noahmpdrv_401.F90 code expects the input precipitation to be an amount.  Within this module, it then converts the precipitation back to a rate, and then calls module_sf_noahmplsm_401.F90.

I'm attaching some slides that explains the differences in the routines between MP-3.6 and MP-4.0.1:

[Noah-MP-4.0.1_code-diffs-MP-3.6.pdf](https://github.com/NASA-LIS/LISF/files/6478581/Noah-MP-4.0.1_code-diffs-MP-3.6.pdf)

Bottom line - the comments are wrong, and should be replaced by:

   296             ! prcp: total precipitation (rainfall+snowfall)
   297             ! Both Noah-MP-3.6 and Noah-MP-4.0.1 require total precipitation as forcing input.
   298             ! In Noah-MP-3.6, the forcing is required to be precipitation rate [kg m-2 sec-1].
   299             ! In Noah-MP-4.0.1, the forcing is required to be precipitation amount [kg m-2].

**Comment by TimLahmers** [comment]
@dmocko Thanks for the update and for fixing the issue with precipitation units.

**Comment by dmocko** [comment]
Thanks @jvgeiger for the assist to get my commit into this PR.

I am fine with this now being merged into the public-7.3 branch, and then put into master.

**Review by jvgeiger — APPROVED** [review]

**Comment by TimLahmers** [comment]
Thanks @jvgeiger @dmocko Which additional steps are needed to merge this into master? Do you need me to do any additional testing?
